vllm/core/evictor_v2.py

```diff
@@ -85,32 +89,44 @@ def evict(self) -> Tuple[int, int]:
         if len(self.free_table) == 0:
             raise ValueError("No usable cache memory left")
 
-        evicted_block = next(iter(self.free_table.values()))
-        evicted_block_id = next(iter(self.free_table.keys()))
-        # The blocks with the lowest timestamps should be placed consecutively
-        # at the start of OrderedDict. Loop through all these blocks to
-        # find the one with maximum number of hashed tokens.
-        for _id, block in self.free_table.items():
-            if evicted_block.last_accessed < block.last_accessed:
-                break
-            if (evicted_block.last_accessed == block.last_accessed and
-                    evicted_block.num_hashed_tokens < block.num_hashed_tokens):
-                evicted_block = block
-                evicted_block_id = _id
-
-        self.free_table.pop(evicted_block_id)
+        while self.priority_queue:
+            # Lazy deletion algorithm is applied.
+            last_accessed, _, content_hash, block_id = heapq.heappop(
+                self.priority_queue)
+            if (block_id in self.free_table and
+                    self.free_table[block_id].last_accessed == last_accessed):
+                self.free_table.pop(block_id)
+                return block_id, content_hash
 
-        return evicted_block_id, evicted_block.content_hash
+        raise ValueError("No usable cache memory left")
 
     def add(self, block_id: int, content_hash: int, num_hashed_tokens: int,
             last_accessed: float):
         self.free_table[block_id] = BlockMetaData(content_hash,
                                                   num_hashed_tokens,
                                                   last_accessed)
+        heapq.heappush(
+            self.priority_queue,
+            (last_accessed, -num_hashed_tokens, content_hash, block_id))
+        self._cleanup_if_necessary()
```

Comment:
Why it was necessary to delay the cleanup? Did you find it to be too slow?

Reply:
The reason I applied lazy deletion and event triggered cleanup is that deleting outdated blocks from the heap is `O(log n)`. Thus, I skip and pop outdated blocks by checking the `free_table` in eviction operation, and only clean up the priority queue when it consumes too much memory with outdated blocks.

Since cleanup itself is O(n log n), calling the cleanup function every time would make the system too slow.

Reply:
The ideal scenario is when the cleanup function is not needed, as outdated blocks are naturally removed during the eviction operation.

Reply:
@alexm-neuralmagic, thanks to your comment, I fixed the data type mistake and optimized the performance of the cleanup operation.

I used only the `free_table` and `heapify` to create a new priority queue, achieving `O(n)` complexity.